crates/opencascade/src/primitives.rs

```diff
@@ -674,6 +834,30 @@ impl Shape {
         self.inner = TopoDS_Shape_to_owned(filleted_shape);
     }
 
+    pub fn chamfer_edges<'a>(&mut self, distance: f64, edges: impl IntoIterator<Item = &'a Edge>) {
+        let mut make_chamfer = BRepFilletAPI_MakeChamfer_ctor(&self.inner);
+
+        for edge in edges.into_iter() {
+            make_chamfer.pin_mut().add_edge(distance, &edge.inner);
+        }
+
+        let chamfered_shape = make_chamfer.pin_mut().Shape();
+
+        self.inner = TopoDS_Shape_to_owned(chamfered_shape);
+    }
+
+    /// Performs fillet of `radius` on all edges of the shape
+    pub fn fillet(&mut self, radius: f64) {
+        let all_edges = self.edges().collect::<Vec<_>>();
+        self.fillet_edges(radius, &all_edges);
```

Comment:
@strohel I was running into this issue as well and I forget the best way to address this in Rust:

`edges()` returns an Iterator of owned Edges, but `chamfer_edges()` accepts an Iterator of borrowed Edges. Is there a better way to structure the function signatures so it can accept code like this?

```rust
self.chamfer_edges(distance, self.edges());
```

Basically trying to avoid the intermediate `Vec`.

Reply:
@bschwind yeah this one is tricky in Rust.

Basically, the `Iterator` trait in `std` is written in a way that it is not possible for the yielded items to borrow from the _iterator_ itself [1]. Think of an iterator that generates the items on the fly. It is easy to create a reference for each single item yielded - but not all of them simultaneously: one would need to store them somewhere (that's `collect()`).

No solution on the iterator side, but multiple possible solutions on the `chamfer_edges()` signature side. I think the canonical solution would be to make change its signature from accepting `impl IntoIterator<Item = &'a Edge>` to accepting `impl IntoIterator<Item = T>` where `T` implements some `EdgeTrait`. `EdgeTrait` would then be implemented for both `Edge` and `&Edge`. This compiles:
```diff
diff --git a/crates/opencascade/src/primitives.rs b/crates/opencascade/src/primitives.rs
index 941a1d4..c34133d 100644
--- a/crates/opencascade/src/primitives.rs
+++ b/crates/opencascade/src/primitives.rs
@@ -180,6 +180,22 @@ impl Edge {
     pub fn tangent_arc(_p1: DVec3, _tangent: DVec3, _p3: DVec3) {}
 }
 
+pub trait EdgeTrait {
+    fn inner(&self) -> &TopoDS_Edge;
+}
+
+impl EdgeTrait for Edge {
+    fn inner(&self) -> &TopoDS_Edge {
+        &self.inner
+    }
+}
+
+impl EdgeTrait for &Edge {
+    fn inner(&self) -> &TopoDS_Edge {
+        &self.inner
+    }
+}
+
 pub struct ApproximationSegmentIterator {
     count: usize,
     approximator: UniquePtr<GCPnts_TangentialDeflection>,
@@ -822,11 +838,11 @@ impl Shape {
         self.inner = TopoDS_Shape_to_owned(chamfered_shape);
     }
 
-    pub fn fillet_edges<'a>(&mut self, radius: f64, edges: impl IntoIterator<Item = &'a Edge>) {
+    pub fn fillet_edges<T: EdgeTrait>(&mut self, radius: f64, edges: impl IntoIterator<Item = T>) {
         let mut make_fillet = BRepFilletAPI_MakeFillet_ctor(&self.inner);
 
         for edge in edges.into_iter() {
-            make_fillet.pin_mut().add_edge(radius, &edge.inner);
+            make_fillet.pin_mut().add_edge(radius, edge.inner());
         }
 
         let filleted_shape = make_fillet.pin_mut().Shape();
@@ -849,7 +865,11 @@ impl Shape {
     /// Performs fillet of `radius` on all edges of the shape
     pub fn fillet(&mut self, radius: f64) {
         let all_edges = self.edges().collect::<Vec<_>>();
+        // operate on iterator of `&Edge`
         self.fillet_edges(radius, &all_edges);
+
+        // but also on iterator of `Edge`.
+        self.fillet_edges(radius, self.edges());
     }
 
     /// Performs chamfer of `distance` on all edges of the shape
```

But it would be cumbersome to introduce something like `EdgeTrait` for each primitive type, and it also exposes internal details.

Fortunately there's a generic trait that we can use for the purpose: [`AsRef`](https://doc.rust-lang.org/std/convert/trait.AsRef.html). With that it becomes (compile-tested):
```diff
diff --git a/crates/opencascade/src/primitives.rs b/crates/opencascade/src/primitives.rs
index 941a1d4..6278126 100644
--- a/crates/opencascade/src/primitives.rs
+++ b/crates/opencascade/src/primitives.rs
@@ -180,6 +180,14 @@ impl Edge {
     pub fn tangent_arc(_p1: DVec3, _tangent: DVec3, _p3: DVec3) {}
 }
 
+// We need to provide this trivial impl: `AsRef` doesn't have this as a blanket implementation for
+// technical reasons: https://doc.rust-lang.org/std/convert/trait.AsRef.html#reflexivity
+impl AsRef<Edge> for Edge {
+    fn as_ref(&self) -> &Edge {
+        self
+    }
+}
+
 pub struct ApproximationSegmentIterator {
     count: usize,
     approximator: UniquePtr<GCPnts_TangentialDeflection>,
@@ -822,11 +830,15 @@ impl Shape {
         self.inner = TopoDS_Shape_to_owned(chamfered_shape);
     }
 
-    pub fn fillet_edges<'a>(&mut self, radius: f64, edges: impl IntoIterator<Item = &'a Edge>) {
+    pub fn fillet_edges<T: AsRef<Edge>>(
+        &mut self,
+        radius: f64,
+        edges: impl IntoIterator<Item = T>,
+    ) {
         let mut make_fillet = BRepFilletAPI_MakeFillet_ctor(&self.inner);
 
         for edge in edges.into_iter() {
-            make_fillet.pin_mut().add_edge(radius, &edge.inner);
+            make_fillet.pin_mut().add_edge(radius, &edge.as_ref().inner);
         }
 
         let filleted_shape = make_fillet.pin_mut().Shape();
@@ -849,7 +861,11 @@ impl Shape {
     /// Performs fillet of `radius` on all edges of the shape
     pub fn fillet(&mut self, radius: f64) {
         let all_edges = self.edges().collect::<Vec<_>>();
+        // operate on iterator of `&Edge`
         self.fillet_edges(radius, &all_edges);
+
+        // but also on iterator of `Edge`.
+        self.fillet_edges(radius, self.edges());
     }
 
     /// Performs chamfer of `distance` on all edges of the shape
```

That's rather neat, so I lean towards going this route.

**TL;DR:** If you need to accept both `T` and `&T` (especially with iterators and other not-simple-to-convert cases), accept `AsRef<T>`. But be sure that `impl AsRef<T> for T` exists - `stdlib` doesn't provide it.

[1] But they can borrow from some other value transitively - i.e. if the iterator already borrows from something (has a lifetime parameter), the yielded items can have that lifetime parameter. That's why `my_vec.iter()` which yields `&T` can work.

Reply:
The `AsRef` solution is perfect, thanks! I actually tried `Iterator<Item = AsRef<Edge>>` last night and for some reason my brain completely forgot about generic bounds.

Reply:
I tried nested `impl` keywords `edge: impl IntoIterator<Item: impl AsRef<Edge>>`, also no avail. :shrug: